include/graphblas/reference/blas3.hpp

```diff
@@ -1111,92 +1118,545 @@ namespace grb {
-				for( size_t j = 1; j < n; ++j ) {
-					CCS_raw.col_start[ j + 1 ] += CCS_raw.col_start[ j ];
+				if( !crs_only ) {
+					assert( CCS_raw.col_start[ 0 ] == 0 );
+					for( size_t j = 1; j < n; ++j ) {
+						CCS_raw.col_start[ j + 1 ] += CCS_raw.col_start[ j ];
+					}
+					assert( CCS_raw.col_start[ n ] == nzc );
 				}
-				assert( CCS_raw.col_start[ n ] == nzc );
 
-				// set C_col_index to all zero
+				// do computations
+				nzc = 0;
+				CRS_raw.col_start[ 0 ] = 0;
+				for( size_t i = 0; i < m; ++i ) {
+					coors1.clear();
+
 #ifdef _H_GRB_REFERENCE_OMP_BLAS3
-				#pragma omp parallel
-				{
-					size_t start, end;
-					config::OMP::localRange( start, end, 0, n );
+					#pragma omp parallel default(none) \
+							shared(coors1, valbuf) \
+							firstprivate(i, A_raw, dummy_identity)
+#endif
+					{
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+						auto local_update = coors1.EMPTY_UPDATE();
+						const size_t maxAsyncAssigns = coors1.maxAsyncAssigns();
+						size_t assigns = 0;
+						#pragma omp for simd schedule( dynamic, config::CACHE_LINE_SIZE::value() ) nowait
+#endif
+						for( size_t k = A_raw.col_start[ i ]; k < A_raw.col_start[ i + 1 ]; ++k ) {
+							const size_t k_col = A_raw.row_index[ k ];
+
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+							if( !coors1.asyncAssign( k_col, local_update ) ) {
+								assignValue( valbuf, k_col , A_raw.getValue( k, dummy_identity ) );
+								if( ++assigns == maxAsyncAssigns ) {
+									coors1.joinUpdate( local_update );
+									assigns = 0;
+								}
+							}
 #else
-					const size_t start = 0;
-					const size_t end = n;
+							if( !coors1.assign( k_col ) ) {
+								assignValue( valbuf, k_col, A_raw.getValue( k, dummy_identity ) );
+							}
 #endif
-					for( size_t j = start; j < end; ++j ) {
-						C_col_index[ j ] = 0;
-					}
+						}
 #ifdef _H_GRB_REFERENCE_OMP_BLAS3
+						while( !coors1.joinUpdate( local_update ) ) {}
+#endif
+					}
+
+					for( size_t l = B_raw.col_start[ i ]; l < B_raw.col_start[ i + 1 ]; ++l ) {
+						const size_t j = B_raw.row_index[ l ];
+						if( !coors1.assigned( j ) ) { // Union case: ignored
+							continue;
+						}
+
+						const auto valbuf_value_before = valbuf[ j ];
+						OutputType result_value;
+						(void)grb::apply( result_value, valbuf_value_before, B_raw.getValue( l, dummy_identity ), oper );
+
+						// update CRS
+						CRS_raw.row_index[ nzc ] = j;
+						CRS_raw.setValue( nzc, result_value );
+
+						// update CCS
+						if( !crs_only ) {
+							++C_col_index[ j ];
+							const size_t CCS_index =  CCS_raw.col_start[ j+1 ] - C_col_index[ j ];
+							CCS_raw.row_index[ CCS_index ] = i;
+							CCS_raw.setValue( CCS_index, result_value );
+						}
+
+						// update count
+						(void)++nzc;
+					}
+
+					CRS_raw.col_start[ i + 1 ] = nzc;
+
+				}
+
+
+#ifdef _DEBUG
+				std::cout << "CRS_raw.col_start = [ ";
+				for( size_t j = 0; j <= m; ++j )
+					std::cout << CRS_raw.col_start[ j ] << " ";
```

Comment:
curly brackets even if the inner code block is a single statement (elsewhere below also)